packages/apollo-server-plugin-full-query-cache/src/ApolloServerPluginFullQueryCache.ts

```diff
@@ -0,0 +1,273 @@
+  cache?: KeyValueCache;
+
+  // Define this hook if you're setting any cache hints with scope PRIVATE.
+  // This should return a session ID if the user is "logged in", or null if
+  // there is no "logged in" user.
+  //
+  // If a cachable response has any PRIVATE nodes, then:
+  // - If this hook is not defined, a warning will be logged and it will not be cached.
+  // - Else if this hook returns null, it will not be cached.
+  // - Else it will be cached under a cache key tagged with the session ID and
+  //   mode "private".
+  //
+  // If a cachable response has no PRIVATE nodes, then:
+  // - If this hook is not defined or returns null, it will be cached under a cache
+  //   key tagged with the mode "no session".
+  // - Else it will be cached under a cache key tagged with the mode
+  //   "authenticated public".
+  //
+  // When reading from the cache:
+  // - If this hook is not defined or returns null, look in the cache under a cache
+  //   key tagged with the mode "no session".
+  // - Else look in the cache under a cache key tagged with the session ID and the
+  //   mode "private". If no response is found in the cache, then look under a cache
+  //   key tagged with the mode "authenticated public".
+  //
+  // This allows the cache to provide different "public" results to anonymous
+  // users and logged in users ("no session" vs "authenticated public").
+  //
+  // A common implementation of this hook would be to look in
+  // requestContext.request.http.headers for a specific authentication header or
+  // cookie.
+  //
+  // This hook may return a promise because, for example, you might need to
+  // validate a cookie against an external service.
+  sessionId?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<string | null>;
+
+  // Define this hook if you want the cache key to vary based on some aspect of
+  // the request other than the query document, operation name, variables, and
+  // session ID. For example, responses that include translatable text may want
+  // to return a string derived from
+  // requestContext.request.http.headers.get('Accept-Language'). The data may
+  // be anything that can be JSON-stringified.
+  extraCacheKeyData?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<any>;
+
+  // If this hook is defined and returns false, the plugin will not read
+  // responses from the cache.
+  readFromCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<boolean>;
+
+  // If this hook is defined and returns false, the plugin will not write the
+  // response to the cache.
+  writeToCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<boolean>;
+}
+
+enum SessionMode {
+  NoSession,
+  Private,
+  AuthenticatedPublic,
+}
+
+function sha(s: string) {
+  return createHash('sha256')
+    .update(s)
+    .digest('hex');
+}
+
+interface BaseCacheKey {
+  document: string;
```

Comment:
A `document` property containing the document text seems confusing, because we use `document` on the request context to refer to the parsed `DocumentNode`.